### PR TITLE
Add assets block

### PR DESCRIPTION
**Pull request by BeksOmega**
### Description

Depends #25

Adds an assets block which can be passed to blocks like `Image.Picture` which require an the name of an asset. Adds an annotation for associating asset blocks with parameters.

I also added filtering by file type, but I'm not sure that I like the behavior. Currently it removes non-matching file names from the dropdown, so it might be better if it was a warning instead. But it feels like the complications to the code outweight the benefits of the user.

Luckily I kept all of that in separate commits, so it'll be easy to revert if desired.

**Important Note**: This PR doesn't include any upgraders because it's not in the right branch path. They can always be added later.

### Testing

1) Uploaded a 2 files to the Media panel.
2) Accessed the asset block via typeblock.
3) Deleted the selected asset via the Media panel.
4) Observed how the block was correctly marked as a badBlock().
BadBlock

5) Reuploaded the file.
6) Observed how the block was unmarked.
7) Deleted both files from the Media panel.
8) Observed how the block correctly displayed that there were no assets available.
NoAssets

9) Reuploaded both files.
10) Observed how the dropdown was correctly displayed.

### CODE REVIEW

**Comment by BeksOmega** [comment]
Closing this until it's more likely to be reviewed.

**Comment by BeksOmega** [comment]
Created some general [helper block documentation](https://docs.google.com/document/d/1D5K_1JwGzcRFSYFq4n9J8ESK2nJJ_TsNwBtrX8w6-9g/edit?usp=sharing).

**Comment by BeksOmega** [comment]
I'll update this with the equivalent of the screen changes tomorrow :D

File: appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/BlocklyPanel.java [hunk 1]
```diff
@@ -12,7 +12,12 @@
 import com.google.appinventor.client.ErrorReporter;
 import com.google.appinventor.client.Ode;
 import com.google.appinventor.client.TopToolbar;
+import com.google.appinventor.client.explorer.project.Project;
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/BlocklyPanel.java:15]: Lots of imports added here, but they don't appear to be used elsewhere in this file.

File: appinventor/blocklyeditor/src/blocks/helpers.js [hunk 2]
```diff
@@ -134,6 +134,12 @@ Blockly.Blocks['helpers_screen_names'] = {
     this.setOutput(true, utils.YailTypeToBlocklyType('text', utils.OUTPUT));
     this.appendDummyInput()
         .appendField(dropdown, 'SCREEN');
+
+    this.workspace.addChangeListener(function(e) {
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/helpers.js]: I'm confused by why this is here, given that it affects screens. What's the motivation for this addition? It also appears redundant.

Comment (ewpatton) [reply appinventor/blocklyeditor/src/blocks/helpers.js]: You should also remove the event listener when the block is disposed, otherwise we will just get a pile of these event listeners for every block added/deleted.

File: appinventor/blocklyeditor/src/blocks/helpers.js [hunk 3]
```diff
@@ -176,3 +182,125 @@ Blockly.Blocks['helpers_screen_names'] = {
...
+    var utils = Blockly.Blocks.Utilities;
+
+    this.setColour(Blockly.COLOUR_HELPERS);
+
+    this.setOutput(true, utils.YailTypeToBlocklyType('text', utils.OUTPUT));
+    this.appendDummyInput('INPUT')
+
+    this.addField();
+
+    this.workspace.addChangeListener(function(e) {
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/helpers.js]: See my comment about the screen version of this code.

File: appinventor/components/src/com/google/appinventor/components/annotations/Asset.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * Annotation used to mark a parameter as accepting an asset string. This can be used to upgrade old
+ * components (ie components that were released before this was released). It can also be included
+ * in new components.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.PARAMETER})
+public @interface Asset {
+    public String[] value() default {};
+}
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/annotations/Asset.java]: Missing new line at end of file.

File: appinventor/components/src/com/google/appinventor/components/common/YaVersion.java [hunk 5]
```diff
@@ -514,7 +514,7 @@ private YaVersion() {
   // - BLOCKS_LANGUAGE_VERSION was incremented to 32
 
   // TODO: Bump this before merge. Include notes about all upgraded components.
-  public static final int YOUNG_ANDROID_VERSION = 207;
+  public static final int YOUNG_ANDROID_VERSION = 208;
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/common/YaVersion.java]: I think you didn't mean to bump this based on your TODO comment.

File: appinventor/blocklyeditor/src/blocks/utilities.js [hunk 6]
```diff
@@ -95,6 +97,17 @@ Blockly.Blocks.Utilities.optionListKeyToBlocklyType = function(key, block) {
   return optionList.className + 'Enum';
 }
 
+/**
+ * Returns a filter array associated with the given key. This can be added to
+ * the connections type check. It causes any asset blocks attached to that
+ * connection to filter their dropdowns.
+ * @param {number=} key The key associated with a filter.
+ * @param {!Blockly.Block} block The block to apply the filter to.
+ */
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/utilities.js:108]: Add the `@return` here so we know the type of this thing.

File: appinventor/blocklyeditor/src/generators/yail/helpers.js [hunk 7]
```diff
@@ -34,3 +34,11 @@ Blockly.Yail['helpers_screen_names'] = function() {
   var value = Blockly.Yail.quote_(this.getFieldValue('SCREEN'));
   return [value, Blockly.Yail.ORDER_ATOMIC];
 }
+
+Blockly.Yail['helpers_assets'] = function() {
+  var field = this.getField('ASSET');
+  if (!field) {
+    return [Blockly.Yail.quote_(' '), Blockly.Yail.ORDER_ATOMIC];
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/generators/yail/helpers.js]: Why are we using a string with a space versus an empty string here?

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java [hunk 8]
```diff
@@ -1944,6 +1956,39 @@ private Option elementToOption(Element field, String value) {
...
+   *
+   * @param elem the Element which represents a function (for return types) or a parameter.
+   * @param type the TypeMirror representing the type of that element.
+   * @return the associated helper key if the element has an @Asset annotation.
+   */
+  private HelperKey hasAssetsHelper(Element elem, TypeMirror type) {
+    for (AnnotationMirror mirror : elem.getAnnotationMirrors()) {
+      if (mirror.getAnnotationType().asElement().getSimpleName().contentEquals("Asset")) {
+        int index = 0;  // Index 0 is the empty filter.
+        for(Entry<? extends ExecutableElement, ? extends AnnotationValue> entry:
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: Space between `for` and `(`, and a space before the `:`

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java [hunk 9]
```diff
@@ -1944,6 +1956,39 @@ private Option elementToOption(Element field, String value) {
...
+            mirror.getElementValues().entrySet()) {
+          // Make sure we are looking at the value attribute.
+          if (!entry.getKey().getSimpleName().contentEquals("value")) {
+            continue;
+          }
+          List<AnnotationValue> values = (List<AnnotationValue>)entry.getValue().getValue();
+          List<String> filter = new ArrayList<String>();
+          for (AnnotationValue v : values) {
+            filter.add((String)v.getValue());
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: I think we should also toLowerCase and sort the filter values. Otherwise, the following annotations are different:

1. `@Assets({".ext1", ".ext2"})`
2. `@Assets({".ext2", ".ext1"})`
3. `@Assets({".EXT1", ".EXT2"))`
4. `@Assets({".EXT2", ".EXT1"))`

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java [hunk 10]
```diff
@@ -736,15 +743,15 @@ protected String getRwString() {
    * blocks for the current UI implementation in the blocks editor. In the future helper keys could
    * be used to define other kinds of helper blocks like range blocks, default color blocks, etc.
    */
-  protected final class HelperKey {
+  protected final class HelperKey<T> {
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: If you're making this generic, it would be good to also make uses of it generic. Right now all of your functions return `HelperKey` and should either be `HelperKey<?>` or specify a specific type in the `<>`.

File: appinventor/blocklyeditor/src/component_database.js [hunk 11]
```diff
@@ -683,6 +732,15 @@ Blockly.ComponentDatabase.prototype.forEachOptionList = function(callback) {
   goog.object.forEach(this.optionLists_, callback);
 }
 
+/**
+ * Returns the asset filter associated with the given key, or undefined.
+ * @param {number} key The key associated with a given filter.
+ * @return {Array<string>=}
+ */
+Blockly.ComponentDatabase.prototype.getFilter = function(key) {
+  return this.filters_[key];
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/component_database.js:743]: If key is out of bounds we should probably return `[]` instead of causing an error to be thrown.

File: appinventor/blocklyeditor/src/component_database.js [hunk 12]
```diff
@@ -504,6 +514,45 @@ Blockly.ComponentDatabase.prototype.processOption = function(option) {
...
+Blockly.ComponentDatabase.prototype.processAssetHelper = function(data) {
+  var filter = data.filter;
+  if (!filter || filter.length == 0) {
+    return {
+      type: "ASSET",
+      key: null
+    }
+  }
+  // We have to do this because js is very restrictive with array equality.
+  var findIndex = function(acc, cur, idx) {
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/component_database.js]: I think the recommended best practice is to just use `function findIndex(..)` instead of assigning the anonymous function to a variable.

File: appinventor/blocklyeditor/src/component_database.js [hunk 13]
```diff
@@ -504,6 +514,45 @@ Blockly.ComponentDatabase.prototype.processOption = function(option) {
...
 }
 
+/**
+ * Processes data defining an asset filter (from simple_components.json) and
+ * returns a HelperKey pointing to the filter.
+ * @param {!Object} data The data defining the filter.
+ * @return {!HelperKey} The key associated with the filter.
+ */
+Blockly.ComponentDatabase.prototype.processAssetHelper = function(data) {
+  var filter = data.filter;
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/component_database.js:524]: There is a minor optimization that can be done here if the filter is sorted where you use binary search on the registered filters using the target filter. Since the elements are ordered, if they don't match starting from the first element you can fail quickly and move to the next potential registered element. I'm not sure it would warrant a rewrite this late in the game but you may just want to note it as a TODO.

File: appinventor/blocklyeditor/src/versioning.js [hunk 14]
```diff
@@ -2565,7 +2656,9 @@ Blockly.Versioning.AllUpgradeMaps =
          Blockly.Versioning.makeSetterUseDropdown(
            'Form', 'AlignHorizontal', 'HorizontalAlignment'),
          Blockly.Versioning.makeSetterUseDropdown(
-           'Form', 'AlignVertical', 'VerticalAlignment')]
+           'Form', 'AlignVertical', 'VerticalAlignment'),
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/versioning.js]: Don't forget to update the comment.

**Review by ewpatton — CHANGES_REQUESTED** [review]
I'm a little concerned about the combination of filtering assets by type + the automatic upgrade. In particular, there are other extension types such as ".mov", ".mkv", ".aac", ".heic", ".webm", etc. that aren't accounted for. Further, it's possible that someone can do the following (and it is the reason for the presence of getBitmapDrawableAsync):

```
set Screen1.BackgroundImage to "http://example.com/image.jpg"
```

If the text isn't a valid asset file, we should definitely not convert it to the asset block. This needs to be accounted for in the `tryReplaceBlockWithAssets` method. Although combined with the extension issue either we need to drop that or just not attempt to upgrade old projects and leave them using strings.

File: appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/BlocklyPanel.java (hunk 1 above)
Comment (BeksOmega) [reply appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/BlocklyPanel.java:15]: Thanks for the catch! I used to have a ProjectChangeListener in this file, but I put it in the YaBlocksEditor file instead.

File: appinventor/blocklyeditor/src/blocks/helpers.js (hunk 2 above)
Comment (BeksOmega) [reply appinventor/blocklyeditor/src/blocks/helpers.js]: Yeah this PR has gotten very messy over time... If I were to rewrite it I'd make it much easier to review haha.

So I had a [listener inside the FieldInvalidDropdown](https://github.com/BeksOmega/appinventor-sources/pull/11/files#diff-a61ef7a00c00fadab2852fa9827eebd8L44) but I realized it wasn't necessary for the helpers_dropdown blocks, so I moved the listeners into the screens & assets fields.

> You should also remove the event listener when the block is disposed, otherwise we will just get a pile of these event listeners for every block added/deleted.

Thanks for the catch! Actually these should be assigned to the onchange property. Those get removed automatically.

File: appinventor/blocklyeditor/src/generators/yail/helpers.js (hunk 7 above)
Comment (BeksOmega) [reply appinventor/blocklyeditor/src/generators/yail/helpers.js]: Good point!

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java (hunk 10 above)
Comment (BeksOmega) [reply appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: Is there a better way to structure this than making it generic? It feels like I'm trying to bring my untyped JS brain to Java, and there may be a better pattern for this. Any opinions?

Also fixed :D

File: appinventor/blocklyeditor/src/component_database.js (hunk 11 above)
Comment (BeksOmega) [reply appinventor/blocklyeditor/src/component_database.js:743]: If the key is out of bounds it will return undefined. That works in all situations where this is called from :D

File: appinventor/blocklyeditor/src/component_database.js (hunk 13 above)
Comment (BeksOmega) [reply appinventor/blocklyeditor/src/component_database.js:524]: I think .every does fail quickly, so now that we are sorting the filter elements inside the ComponentProcessor we should get that benefit :D

But I think that you're advising that we should sort the filters in this.filters_ in addition to the elements in each filter. I'll add a TODO for that.

**Comment by BeksOmega** [comment]
> I'm a little concerned about the combination of filtering assets by type + the automatic upgrade. In particular, there are other extension types such as ".mov", ".mkv", ".aac", ".heic", ".webm", etc. that aren't accounted for. Further, it's possible that someone can do the following (and it is the reason for the presence of getBitmapDrawableAsync):
> 
> ```
> set Screen1.BackgroundImage to "http://example.com/image.jpg"
> ```
> 
> If the text isn't a valid asset file, we should definitely not convert it to the asset block. This needs to be accounted for in the `tryReplaceBlockWithAssets` method. Although combined with the extension issue either we need to drop that or just not attempt to upgrade old projects and leave them using strings.

So it seems like we have two things to decide:
1) Do we keep filtering, or do we remove it?
2) Do we keep upgrading & fix it for async calls, or do we remove upgrading.

Personally I think we should remove the filtering and fix the upgrading. I lean towards removing the filtering and fixing the upgrading, but It's not a strong feeling.

**Comment by ewpatton** [comment]
@BeksOmega 

> 1. Do we keep filtering, or do we remove it?

I think we should disable the filtering for now. We can leave it in but turn it off for the time being. I'm mostly concerned that we can't really know all the supported formats and we don't want to jump the gun here.

> 2. Do we keep upgrading & fit it for async calls, or do we remove upgrading.

I think we can keep the upgrading and just check the string against the list of assets. When the blocks file loads, we will know all of the assets available in the project.

**Comment by BeksOmega** [comment]
> I think we should disable the filtering for now. We can leave it in but turn it off for the time being. I'm mostly concerned that we can't really know all the supported formats and we don't want to jump the gun here.

Sounds good to me! Should we still make it available to extension developers? Or should I disable it for everyone and remove the docs on it?

> I think we can keep the upgrading and just check the string against the list of assets. When the blocks file loads, we will know all of the assets available in the project.

Coolio, can do :D

**Comment by ewpatton** [comment]
@BeksOmega Is this all set for another review?

**Comment by BeksOmega** [comment]
@ewpatton Yep should be good for another look =)

Note that I didn't mess with filtering yet though, still waiting for a response on:
> Should we still make [filtering] available to extension developers? Or should I disable it for everyone and remove the docs on it?

**Comment by ewpatton** [comment]
> Should we still make [filtering] available to extension developers? Or should I disable it for everyone and remove the docs on it?

Sorry I missed that one. I think it's okay to leave it in and let extension developers make the call for themselves. Like I said, I'm hesitant on us making the call but it's possible that extensions have more specific requirements. For example, an extension that only displays PDFs can strongly assert that .pdf files are acceptable and nothing else. I'm more concerned about us clamping down prematurely on core components where developers are used to having a broader range of options.

**Comment by BeksOmega** [comment]
> Sorry I missed that one. I think it's okay to leave it in and let extension developers make the call for themselves. Like I said, I'm hesitant on us making the call but it's possible that extensions have more specific requirements. For example, an extension that only displays PDFs can strongly assert that .pdf files are acceptable and nothing else. I'm more concerned about us clamping down prematurely on core components where developers are used to having a broader range of options.

Coolio. That was my instinct too, but I wanted to make sure that's what you wanted =)

I've reverted the PR that added filters to built-in components. So this should be all good for a look @ewpatton .

File: appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/YaBlocksEditor.java [hunk 15]
```diff
@@ -734,24 +729,31 @@ public void makeActiveWorkspace() {
   @Override
   public void onProjectLoaded(Project project) {
     for (ProjectNode node : project.getRootNode().getAllSourceNodes()) {
-      if (node instanceof YoungAndroidSourceNode) {
-        addScreen(((YoungAndroidSourceNode)node).getFormName());
-      }
+      blocksArea.addScreen(((YoungAndroidSourceNode)node).getFormName());
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/YaBlocksEditor.java]: You may want to filter and only do this if the node is a YoungAndroidFormNode. Otherwise, you'll call this up to three times for the .scm, .bky, and .yail files for each screen.

Comment (ewpatton) [reply appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/YaBlocksEditor.java]: Minor code style: space between `(YoungAndroidSourceNode)` and `node` in the cast.

File: appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/YaBlocksEditor.java [hunk 16]
```diff
@@ -734,24 +729,31 @@ public void makeActiveWorkspace() {
...
+    for (ProjectNode node : assetsFolder.getChildren()) {
+      blocksArea.addAsset(((YoungAndroidAssetNode)node).getName());
     }
   }
 
   @Override
   public void onProjectNodeAdded(Project project, ProjectNode node) {
     if (node instanceof YoungAndroidSourceNode) {
-      addScreen(((YoungAndroidSourceNode)node).getFormName());
+      blocksArea.addScreen(((YoungAndroidSourceNode)node).getFormName());
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/YaBlocksEditor.java]: See earlier comments about node type and styling.

File: appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/YaBlocksEditor.java [hunk 17]
```diff
@@ -734,24 +729,31 @@ public void makeActiveWorkspace() {
...
+      blocksArea.addAsset(((YoungAndroidAssetNode)node).getName());
     }
   }
 
 
   @Override
   public void onProjectNodeRemoved(Project project, ProjectNode node) {
     if (node instanceof YoungAndroidSourceNode) {
-      removeScreen(((YoungAndroidSourceNode)node).getFormName());
+      blocksArea.removeScreen(((YoungAndroidSourceNode)node).getFormName());
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/YaBlocksEditor.java]: See earlier comments about node type and styling.

File: appinventor/blocklyeditor/src/component_database.js (hunk 13 above)
Comment (ewpatton) [reply appinventor/blocklyeditor/src/component_database.js:524]: You should change the return type of `Blockly.Blocks.Utilities.assetKeyToBlocklyType` to match this function then as it is the only place that it is called from. This return allows for undefined but the caller does not.

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java [hunk 18]
```diff
@@ -209,6 +209,13 @@
    */
   protected final Map<String, OptionList> optionLists = Maps.newTreeMap();
 
+  /**
+   * A list of asset filters, which are in fact lists of strings. Initialized with an empty filter.
+   */
+  protected final List<List<String>> filters = new ArrayList<List<String>>() {{
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: Rather than create an anonymous inner type here subclassing ArrayList, you could instead break the initialization into a separate `static` block:

```java
protected final List<List<String>> filters;

static {
  filters = new ArrayList<String>();
  filters.add(new ArrayList<String>());
}
```

IMO it's more understandable to most people than the double-curly notation.

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java (hunk 10 above)
Comment (ewpatton) [reply appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: Given how it is used in the code, it seems like there are two approaches you could take.

Option 1: Don't use generics. Make the key type be `Object` and continue to cast as you've done where `getKey` is used. That's effectively what happens when you specify the generic as `<?>` or leave it unspecified.

Option 2: Don't make this final and create two new subclasses that specify the key type:

```java
protected class OptionHelperKey extends HelperKey<String> {
  protected OptionHelperKey(String key) {
    super(HelperType.OPTION_LIST, key);
  }
}

protected class AssetHelperKey extends HelperKey<Integer> {
  protected AssetHelperKey(Integer key) {
    super(HelperType.ASSET, key);
  }
}
```

and then in ComponentDescriptorGenerator in your switch where you handle the helper type you can cast to either OptionHelperKey or AssetHelperKey.

Either option is fine because the cast that occurs at runtime will fail if for whatever reason the type is wrong. Generics are really intended for compile time type safety checks, but we're mixing the different types of key helpers into a single data structure.

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java [hunk 19]
```diff
@@ -1944,6 +1956,40 @@ private Option elementToOption(Element field, String value) {
...
   }
 
+  /**
+   * Returns the associated helper key if the element has an @Asset annotation. Null otherwise.
+   *
+   * @param elem the Element which represents a function (for return types) or a parameter.
+   * @param type the TypeMirror representing the type of that element.
+   * @return the associated helper key if the element has an @Asset annotation.
+   */
+  private HelperKey<Integer> hasAssetsHelper(Element elem, TypeMirror type) {
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: I'm curious how this approach will work with extensions. If you and I both create an extension with a filter, can't we end up with the same index? When we load these two extensions into App Inventor, do we end up selecting the same (potentially incorrect) filter?

I recall we decided to drop filtering for now, but you may want to leave a TODO here to think about that.

**Review by ewpatton — CHANGES_REQUESTED** [review]
A few minor thing otherwise LGTM. See the discussion of options for dealing with `HelperKey<T>` and let me know if you have any questions.

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java (hunk 10 above)
Comment (BeksOmega) [reply appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: > Option 2: Don't make this final and create two new subclasses that specify the key type:

Ok this does seem like the more Java way to do things. And I was going to switch it over. But then I was thinking it makes the HelperType kind of confusing when the type is encoded in the actual class, but I don't want to get rid of the HelperType because that makes it more consistent with the js side of the implementation.

So if I was doing this was purely a Java project I'd definitely move to Option 2, but since I want future people to understand how the Java and JS parts match, I'm going to move to Option 1.  I've added the above context to the source.

If you have a different opinion though, I'm happy to do it the other way.

File: appinventor/blocklyeditor/src/component_database.js (hunk 11 above)
Comment (BeksOmega) [reply appinventor/blocklyeditor/src/component_database.js:743]: I think [this](https://github.com/BeksOmega/appinventor-sources/pull/11#discussion_r480084266) was meant to go here:
> You should change the return type of Blockly.Blocks.Utilities.assetKeyToBlocklyType to match this function then as it is the only place that it is called from. This return allows for undefined but the caller does not.

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java (hunk 18 above)
Comment (BeksOmega) [reply appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: I wasn't able to add it inside a `static` block b/c the list isn't static. So I moved it to a constructor for the ComponentProcessor instead.

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java (hunk 19 above)
Comment (BeksOmega) [reply appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: There are no conflict problems wrt extensions because the index in the list in the ComponentProcessor doesn't get encoded in the simple_components.json file. The only info that gets sent to the component_database.js is the actual filter. Eg:
```
   "type": "ASSET",
    "data": {
      "filter": [ ".jpeg" ]
    }
```

And then the component_database de-duplicates all the data and comes up with its own new set of indices.

And because none of the filter data gets saved to XML (all of the behavior is encoded in the connection checks) we don't have any trouble with round tripping using this system.

TL:DR already covered, shouldn't be anything to worry about.

**Comment by BeksOmega** [comment]
@ewpatton if you're cool with moving the filters initialization to the ComponentProcessor constructor I think this is gtg.

**Review by ewpatton — APPROVED** [review]
Yes, I think that's fine.
